packages/story-editor/src/components/canvas/mediaRecordingLayer.js

```diff
@@ -0,0 +1,228 @@
+    getMediaStream: actions.getMediaStream,
+    clearMediaStream: actions.clearMediaStream,
+    setFile: actions.setFile,
+    setMediaBlobUrl: actions.setMediaBlobUrl,
+    resetStream: actions.resetStream,
+  }));
+
+  const streamRef = useRef();
+  const videoRef = useRef();
+
+  const isReady = 'ready' === status;
+  const isFailed = 'failed' === status;
+
+  const isMuted = !hasAudio || isGif;
+
+  const captureImage = useCallback(async () => {
+    if (!streamRef.current) {
+      return;
+    }
+
+    const blob = await getImageFromVideo(streamRef.current);
+    try {
+      setMediaBlobUrl(createBlob(blob));
+    } catch (e) {
+      // Do nothing.
+    }
+
+    const imageFile = blobToFile(
+      blob,
+      `image-capture-${format(new Date(), 'Y-m-d-H-i')}.${PHOTO_FILE_TYPE}`,
+      PHOTO_MIME_TYPE
+    );
+    setFile(imageFile);
+    resetStream();
+  }, [resetStream, setFile, setMediaBlobUrl]);
+
+  const debouncedCaptureImage = useDebouncedCallback(captureImage, 3000);
+
+  const onToggleVideoMode = useCallback(() => {
+    toggleIsGif();
+    if (videoRef.current) {
+      videoRef.current.play().catch(() => {});
+    }
+  }, [toggleIsGif]);
+
+  useEffect(() => {
+    resetStream();
+    getMediaStream();
+    // eslint-disable-next-line react-hooks/exhaustive-deps -- Only want to act on actual input changes.
+  }, [audioInput, videoInput]);
+
+  useEffect(() => {
+    if (isReady) {
+      updateMediaDevices();
+    }
+  }, [isReady, updateMediaDevices]);
+
+  useEffect(() => {
+    // Checking for srcObject avoids flickering due to the stream changing constantly.
+    if (streamRef.current && !streamRef.current.srcObject && liveStream) {
+      streamRef.current.srcObject = liveStream;
+    }
+
+    if (streamRef.current && !liveStream) {
+      streamRef.current.srcObject = null;
+    }
+  }, [liveStream]);
+
+  if (needsPermissions) {
+    return <PermissionsDialog />;
+  }
+
+  if (isFailed) {
+    return <ErrorDialog />;
+  }
+
+  return (
+    <>
+      <LayerWithGrayout aria-label={__('Media Recording layer', 'web-stories')}>
+        <PageTitleArea>
+          <DurationIndicator />
+        </PageTitleArea>
+        <DisplayPageArea withSafezone={false} showOverflow>
+          <Wrapper>
+            {!isImageCapture && (
+              <>
+                {mediaBlobUrl && (
+                  <VideoMode value={!isGif} onChange={onToggleVideoMode} />
+                )}
+                <VideoWrapper>
+                  {mediaBlobUrl && (
+                    <>
+                      {/* eslint-disable-next-line jsx-a11y/media-has-caption -- We don't have tracks for this. */}
+                      <Video
+                        ref={videoRef}
+                        src={mediaBlobUrl}
+                        autoPlay
+                        muted={isMuted}
+                        loop={isGif}
+                        disablePictureInPicture
+                        tabIndex={0}
```

Comment:
I wonder if we should have shared attributes like this. 
```
const videoAttributes = {
   disablePictureInPicture: true,
   autoPlay: true,
   tabIndex: 0, 
   muted: isMuted,
}
```

```
<Video ref={streamRef}  {...videoAttributes} />
```

Reply:
I don't see the need for this. These props are just for this one particular component, not shared with others, and not passed down further in the tree. Moving them away from here to a separate `const` would make it harder to follow.

Reply:
Why are the attributes difference between components? Like tabindex. 

Could these be one component with conditional attributes.

Reply:
One `<Video>` is for displaying your webcam stream before and during recording (the one with the `streamRef`)

The other one (with the `videoRef`) displays the final video after recording and has different props depending on whether it's a GIF or not.
It needs a `tabIndex` so that it can receive keyboard focus, which allows you to still play/pause the video even without any actual video controls being displayed. Otherwise, videos without controls are not interactive.

These two don't really share much as they serve different purposes.

I'm happy to move the only two shared props (`autoPlay` and `disablePictureInPicture`) to the base styled component, but that's the only optimization I can think of here without making the code less readable.